### ISSUE TITLE
Write speed improvements

### ISSUE THREAD

**Issue by marcharding** [winnfsd/winnfsd#22]
**Edit: take a look at the latest comment: https://github.com/winnfsd/winnfsd/issues/22#issuecomment-230125433**

I'm currently working on improving the write speed.

The biggest bottleneck was the very small wsize which is returned in the FSINFO call.
When increasing that to 65536 another problem gets obvious: the buffer size for the sockets is only the windows default of 8192. That small buffer results in endless UDP retransmissions, so i also increased all socket send and receive buffers.

The results are as follows:

Before: 200 kB/s (worstcase) to 2.5 MB/s write throughput
After: ~12.5 MB/s write throughput (max speed of Am79C973, sadly all other network adapters performed worse)

Please try the attached exe [WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/333153/WinNFSd.zip) and take a look at the changes in my rw-speed branch (https://github.com/marcharding/winnfsd/tree/rw-speed).
If you try the attached gem [vagrant-winnfsd-1.2.1ALPHA01.gem.zip](https://github.com/winnfsd/winnfsd/files/333152/vagrant-winnfsd-1.2.1ALPHA01.gem.zip), please see https://github.com/mitchellh/vagrant/issues/7415 as this issue seems to persist for installing local gems, vagrant 1.8.1 works fine.

> vagrant plugin uninstall vagrant-winnfsd
> vagrant plugin install vagrant-winnfsd-1.2.1ALPHA01.gem

When testing, please run winnfsd.exe with -log off, printing all the write calls seems to decrease performance in my tests.

Furthermore, i got the best results using the Am79C973 network adapter in virtualbox (i'm not sure why, but it just worked best)

I'm looking forward to your results.

**Comment by exchentrix** [comment]
Wow! This gives me a huge performance boost:

### Before

![profiler2](https://cloud.githubusercontent.com/assets/1130714/16359119/3788533c-3b29-11e6-9c0c-444d29890014.png)

### After

![profiler3](https://cloud.githubusercontent.com/assets/1130714/16359123/41265e0c-3b29-11e6-95df-9d8756f6b356.png)

Thank you very much for this. :)


Best
Christian

**Comment by eXtreme** [comment]
wow this is really really fast now, I can now even use symfony without moving cache and logs outside the nfs share and the performance is still very good! My test suite runs almost the same time as on my CI server, even with cache and logs folder inside the nfs share.

**Comment by marcharding** [comment]
Hey all,

i added an experimental unstable/async WRITE/COMMIT procedure. Check the attached [WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/344761/WinNFSd.zip) or [vagrant-winnfsd-1.2.1ALPHA02.gem.zip](https://github.com/winnfsd/winnfsd/files/344762/vagrant-winnfsd-1.2.1ALPHA02.gem.zip)

I suggest to use the virtio network adapter with this build, i get about 150 to 200 Mb/s write speed for big files. 

For small file the difference is not that big but try it for yourself.

Also included is a fix for slow READDIR operations with many files in one folder and some improvements in the READLINK procedure.

I'm looking forward to your results.

**Comment by exchentrix** [comment]
This gives me another performance boost:

![profiler](https://cloud.githubusercontent.com/assets/1130714/16544816/44fcea8c-4114-11e6-8d9a-4542fa802a3a.png)

Thanks for this. :+1:

**Comment by eXtreme** [comment]
@marcharding I can't get ALPHA02 to work properly all the time. At first I couldn't get it working, it was crashing at startup, went back to ALPHA01, it worked, reinstalled ALPHA02 and it started working. But I didn't see any gain or losses in performance. I will test it more later this day because now I have to work ;)

**Comment by marcharding** [comment]
I put some (_a lot_) more work into this due to the issues that kept coming up:

- Fixed TCP (i probably recommend this as the default mount option)
- Fixed some async mode crashes
- Some fixes for the symlink procedure

I didn't test this extensively yet, but i'm quite confident that this version will fix some outstanding issues with git/composer (npm still has problems).

So without further ado i present another ALPHA, lets call it ALPHA03: 

**[WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/362588/WinNFSd.zip)** (only the exe this time).

I recommend testing this with `rw,vers=3,tcp,nolock` as your mount options.

**Comment by eXtreme** [comment]
So far, so good, no lockups, no crash, phpunit runs fast, composer too.

**Comment by SirWaddles** [comment]
I was getting lockups on Symfony's assets:install command on the current release version (2.1.1) but Alpha03 seems to have fixed it.

**Comment by fabis94** [comment]
Trying to run ALPHA03, but can't run it on W10 due to 'MSVCR120D.dll' being missed. Tried googling about this and supposedly the exe was built in the Debug configuration? Could that be the case?

**Comment by marcharding** [comment]
All binaries here a created by the ci process on appveyor with the release config. I'll have to check whats going on.

Do the [release binaries](https://github.com/winnfsd/winnfsd/releases) work for you?

In the meantime, try installing http://www.microsoft.com/en-us/download/details.aspx?id=40784 first.

**Comment by fabis94** [comment]
Yes, I can run them and they don't require the DLL. I've also already installed the redistributable. Are you sure you didn't build this last alpha build with a different build configuration?

**Comment by marcharding** [comment]
I'm sure, but someone else had a problem with a failed assertion in that build, maybe that is what triggers your problem. The assertion can probably be removed in total (it seems to be from the original authors).

Does it happen with [this file](https://github.com/winnfsd/winnfsd/files/384857/WinNFSd.zip) too?

**Comment by fabis94** [comment]
No, that file works. Is it the latest ALPHA3 version?

**Comment by bkibilko** [comment]
Unfortunately, I can't install the plugin:
`vagrant plugin install vagrant-winnfsd-1.2.1ALPHA02.gem`

`Installing the 'vagrant-winnfsd-1.2.1ALPHA02.gem' plugin. This can take a few minutes...
Bundler, the underlying system Vagrant uses to install plugins,
reported an error. The error is shown below. These errors are usually
caused by misconfigured plugin installations or transient network
issues. The error from Bundler is:`

`Could not find gem vagrant-winnfsd (= 1.2.1ALPHA02) x86-mingw32 in any of the gem sources listed in your Gemfile or available on this machine.`

`Warning: this Gemfile contains multiple primary sources. Using source more than once without a block is a security risk, and may result in installing unexpected gems. To resolve this warning, use a block to indicate which gems should come from the secondary source. To upgrade this warning to an error, run bundle config disable_multisource true.`

**Comment by marcharding** [comment]
Please upgrade to the latest vagrant version (should be fixed in 1.8.5.), this is a bug in vagrant, see https://github.com/mitchellh/vagrant/pull/7505 and https://github.com/mitchellh/vagrant/issues/7493.

**Comment by bkibilko** [comment]
Hey, it worked ;). I'm going to test it on my project. What is the simplest way to run a benchmark?

**Comment by vmadev** [comment]
@bkibilko Just FYI;

On Windows 10 the only way to get a stable run is with the following mount options;

`      mount_options: ['nolock,vers=3,udp,noatime']` 

In my particular scenario, I am running vagrant up on a vagrant laravel/homestead box with it's associated provisioning scripts but i think the mount options are just a pass through configuration option. I tried the testing option mentioned above; `rw,vers=3,tcp,nolock` and the daemon would simply die after many times less than a minute and I would look at the logs and all the actions would be OK, in other words, no fails...

**Comment by marcharding** [comment]
@vmadev i think you're not using the latest alpha release.

i'm using tcp with windows 10 for more than 8 hours without any problems on quite large symfony 2 project.

Would you mind to try again with [this file](https://github.com/winnfsd/winnfsd/files/384857/WinNFSd.zip)?

**Comment by vmadev** [comment]
ok, just to be sure.. what mount options should work?

**Comment by marcharding** [comment]
Thanks for the update, maybe it has something to do with the insider build you're using.

Does winnfsd crash or just freeze?

Udp has its own quirks (try a `git clone https://github.com/silexphp/Silex.git` and it will freeze with udp).

**Comment by vmadev** [comment]
Well when I was on Windows Pro (Stable Release) on, tcp it would outright crash. The service would just disappear from the task list. 

On this Insider Preview, the daemon appears to stay active and you can even enter commands in the console but there doesnt look like its performing any read and write operations. So it freezes the read and write to disk but you still have limited interaction with the console. Don't know if that helps.

Thanks
Vic

**Comment by exchentrix** [comment]
@marcharding One short question: Can you add support for running multiple boxes with nfs sharing?

**Comment by marcharding** [comment]
@vmadev i'll take a look, i may have screwed up the build.

@exchentrix that should already work, even in the official release (see https://github.com/winnfsd/winnfsd/releases/tag/2.1.0, first bullet point)

**Comment by marcharding** [comment]
@vmadev

> So it freezes the read and write to disk but you still have limited interaction with the console. Don't
know if that helps.

Console as in bash console, e.g. doing a ` ls /your-mount-path/`still works?

Or console as in cmd.exe console (the winnfsd window).

If you're talking about the cmd.exe console, please take a look at https://github.com/winnfsd/winnfsd/issues/15#issuecomment-227469021.

I'm a little bit puzzled, because l just let it run via tcp for a few hours doing a few hundred composer installs (symfony project) without any issues...

Are you using the vagrant-plugin or just the executable (manually starting winnfsd.exe)?

If you're using the plugin, did you really replace the binary that ships with the vagrant plugin with the "alpha" version here?

**Comment by vmadev** [comment]
@marcharding 

I am using the vagrant-plugin 1.2.1 along with the 384857 release of Winnfsd.exe

I usually run vagrant up from an elevated windows powershell using CMDer, I could also use cmd.exe or CMDer's built in bash as admin. But i have run vagrant up from an elevated bash as well. Its mainly npm which i try to keep to powershell. 

Vic

**Comment by marcharding** [comment]
thanks for the prompt reply :+1: 

that should work (same setup here), although i know that npm does not work quite right all the time (see https://github.com/winnfsd/vagrant-winnfsd/issues/96).

i'll see if i can come up with something, but if udp work fine for you, maybe we just leave it at that :wink:

**Comment by vmadev** [comment]
@marcharding Im the one who should be thanking you sir....... 

Thank You.

**Comment by marcharding** [comment]
I'm closing this because i just released 2.2.0 which includes all these changes. 

If any of you still have problems, please open a new issue.

**Comment by LogicByteCraft** [comment]
The version I use is: vagrant-winnfsd (1.4.0)
The command I execute is : tail -f lum-180622.log
The error message is: reverting to polling: Input/output error
I have used an administrator to run
Please help me, thank you@marcharding
